Add PagedMergeSort, a merge sort using O(√n) space

LSchwerdt opened the pull request:
As far as I am aware, there is no stable sorting algorithm in Julia that uses less than O(n) auxiliary space. This PR fixes that, by implementing PagedMergeSort, a merge sort using O(√n) space while achieving about the same speed as the regular MergeSort from Base.Sort.

This is done by using a merge routine that splits the array into blocks/pages of size √n, merges into these blocks, and than rearranges them using a page table. This is illustrated below. The auxiliary space is on the right.

![pagedMerge_130_130](https://user-images.githubusercontent.com/8214073/216924165-c1727446-3968-4175-ad69-542d090b8a84.gif)

The basic idea is laid out [here](https://max-arbuzov.blogspot.com/2021/10/merge-sort-with-osqrtn-auxiliary-memory.html), but I use a different page table and reordering scheme than the author of that post. By using the inverse permutation in the page table (ie. storing the location of the data belonging in block i in the page table at index i), we can follow the permutation cycles during reordering without swapping, copying the correct data into the previously emptied location.

The additional data movement from reordering is almost negligible compared to the merging, and the main merging loop has only one condition (merge until block is full) as opposed to two (merge until either source array runs out), so the performance is about the same as the regular merge sort.

At deeper recursion levels, where the scratch space is big enough, normal merging is used, where one input is copied into the scratch space. When the scratch space is large enough to hold the complete subarray, the input is merged interleaved from both sides, which increases performance for random data, making PagedMergeSort actually faster than MergeSort. (But it decreases performamce for presorted data). Benchmarks are below. They can be replicated by running [this Pluto notebook](https://github.com/LSchwerdt/MiscJulia/blob/b66c633eb5ad856e6d277d08d59172ee6f1b90c3/benchmark_pagedmergesort/benchmark_PagedMergeSort.jl).

If someone needs a sorting algorithm with less than O(n) memory usage, sorting is probably a significant aspect of their whole program, so I also included a multithreaded version of PagedMergeSort.

Checklist before merging
------

- [x] Wait for https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/70 to be merged, and include initial optimizations. This  is necessary to pass the test for sorting floats with NaNs.
- [x] This PR adds a dependency on StaticArrays.jl. Maybe that can be avoided. If not, version compatibility must be checked.
- [x] PagedMerge! uses Base.midpoint, which was added in version 1.4 in Base.Sort and later moved to Base. To ensure compatibility with older Julia version, this should be accounted for. -> Redefined function here for compatibiliy with older Julia versions.
- [x] Threads.@spawn was introduced after Julia 1.0. Since this package aims to be compatible with Julia 1.0, I see two options:
  - Create a second package for multithreaded sorting algorithms with different Julia version requirements.
  - Use the singlethreaded or multithreaded algorithm depending on the Julia version, when calling the multithreaded one. (implemented currently in this PR)
- [x] I am not 100% happy with using a macro (getNextBlock!) instead of a function, but I could not come up with a better way.
- [x] Rename buffer/buf to scratch or to t? -> renamed to scratch
- [x] Improve tests to find errors that were made here
  - Stable sorting with by=Returns(0)
  - Odd input length
- [x] CI only tests the fallback to PagedMergeSort for ThreadedPagedMergeSort. -> deleted ThreadedPagedMergeSort for now
- [x] Rebase on current master.

Some notes
-------

- The basic idea of the algorithm seems not that complicated. Maybe it was already published under a different name? Edit: [I found it](https://link.springer.com/chapter/10.1007/BFb0016253).
- There are stable sorting algorithms that are in-place, but this is easier. And I think allocating 1 MiB to sort 8 GiB is practically in-place for most applications.
- Moving the blocks before merging actually allows for even less data movement, but the required bookeeping is more complicated.
- The two-ended merge can also be applied at all recursion levels to get an even faster merge sort algorithm, that uses auxilary memory the size of the input.
- Splitting the processed data into blocks and rearranging them later can also be used for distributive sorting, see [IPS4o](https://arxiv.org/abs/1705.02257).

Benchmarks
------

TLDR: PagedMergeSort is slightly faster than MergeSort, but slower than the default radix sort on random data. It allocates less than 400 KiB to sort a 1 GiB vector. Both MergeSort and PagedMergeSort are faster for sortperm! than the default ScratchQuicksort when sorting large vectors. The speedup depends on memory latency. ThreadedPagedMergeSort is the fastest algorithm for all tested inputs and machines (but it is the only multithreaded one).

```
Julia Version 1.9.0-beta3
Commit 24204a7344 (2023-01-18 07:20 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 16 × Intel(R) Core(TM) i7-7820X CPU @ 3.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, skylake-avx512)
  Threads: 8 on 16 virtual cores
Environment:
  JULIA_DEPOT_PATH = C:\Users\schwerdt\.julia;D:\Sonstiges\JuliaTest\julia-1.9.0-beta3\local\share\julia;D:\Sonstiges\JuliaTest\julia-1.9.0-beta3\share\julia
  JULIA_IMAGE_THREADS = 1
  JULIA_LOAD_PATH = @;@v#.#;@stdlib
  JULIA_REVISE_WORKER_ONLY = 1

sort! T=UInt64 n=131_072 (1.0 MiB)
default:                      0.003205 seconds (3 allocations: 1.008 MiB)
MergeSort:                    0.009005 seconds (2 allocations: 512.047 KiB)
PagedMergeSort:               0.006894 seconds (2 allocations: 11.625 KiB)
PagedMergeSort, 8 threads:    0.002509 seconds (116 allocations: 103.766 KiB)

sortperm! T=UInt64 n=131_072 (1.0 MiB)
default:                      0.006763 seconds (2 allocations: 1.000 MiB)
MergeSort:                    0.010892 seconds (2 allocations: 512.047 KiB)
PagedMergeSort:               0.010665 seconds (2 allocations: 11.625 KiB)
PagedMergeSort, 8 threads:    0.002941 seconds (117 allocations: 104.031 KiB)

sort! T=UInt64 n=8_388_608 (64.0 MiB)
default:                      0.282875 seconds (3 allocations: 64.008 MiB, 1.28% gc time)
MergeSort:                    0.808292 seconds (2 allocations: 32.000 MiB)
PagedMergeSort:               0.631584 seconds (4 allocations: 90.656 KiB)
PagedMergeSort, 8 threads:    0.148540 seconds (209 allocations: 745.922 KiB)

sortperm! T=UInt64 n=8_388_608 (64.0 MiB)
default:                      4.226174 seconds (2 allocations: 64.000 MiB, 0.01% gc time)
MergeSort:                    1.399869 seconds (2 allocations: 32.000 MiB)
PagedMergeSort:               1.378963 seconds (4 allocations: 90.656 KiB)
PagedMergeSort, 8 threads:    0.453981 seconds (210 allocations: 746.438 KiB)

sort! T=UInt64 n=134_217_728 (1.0 GiB)
default:                      4.341338 seconds (3 allocations: 1.000 GiB, 0.07% gc time)
MergeSort:                   15.697732 seconds (2 allocations: 512.000 MiB, 0.01% gc time)
PagedMergeSort:              12.191662 seconds (4 allocations: 362.219 KiB)
PagedMergeSort, 8 threads:    2.692278 seconds (210 allocations: 2.850 MiB)

sortperm! T=UInt64 n=134_217_728 (1.0 GiB)
default:                    158.911205 seconds (2 allocations: 1.000 GiB, 0.00% gc time)
MergeSort:                   33.823217 seconds (2 allocations: 512.000 MiB, 0.27% gc time)
PagedMergeSort:              34.031642 seconds (4 allocations: 362.219 KiB)
PagedMergeSort, 8 threads:   10.372993 seconds (210 allocations: 2.851 MiB)
```

```
Julia Version 1.9.0-beta3
Commit 24204a7344 (2023-01-18 07:20 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 8 × Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, skylake)
  Threads: 8 on 8 virtual cores
Environment:
  JULIA_DEPOT_PATH = C:\Users\Luke\.julia;C:\Users\Luke\.julia\juliaup\julia-1.9.0-beta3+0.x64.w64.mingw32\local\share\julia;C:\Users\Luke\.julia\juliaup\julia-1.9.0-beta3+0.x64.w64.mingw32\share\julia
  JULIA_IMAGE_THREADS = 1
  JULIA_LOAD_PATH = @;@v#.#;@stdlib
  JULIA_NUM_THREADS = auto
  JULIA_REVISE_WORKER_ONLY = 1

sort! T=UInt64 n=131_072 (1.0 MiB)
default:                      0.010040 seconds (3 allocations: 1.008 MiB)
MergeSort:                    0.015199 seconds (2 allocations: 512.047 KiB)
PagedMergeSort:               0.010011 seconds (2 allocations: 11.625 KiB)
PagedMergeSort, 8 threads:    0.004095 seconds (115 allocations: 103.734 KiB)

sortperm! T=UInt64 n=131_072 (1.0 MiB)
default:                      0.011959 seconds (2 allocations: 1.000 MiB)
MergeSort:                    0.017257 seconds (2 allocations: 512.047 KiB)
PagedMergeSort:               0.018372 seconds (2 allocations: 11.625 KiB)
PagedMergeSort, 8 threads:    0.007372 seconds (117 allocations: 104.031 KiB)

sort! T=UInt64 n=8_388_608 (64.0 MiB)
default:                      0.668358 seconds (3 allocations: 64.008 MiB, 0.57% gc time)
MergeSort:                    0.997630 seconds (2 allocations: 32.000 MiB)
PagedMergeSort:               0.845741 seconds (4 allocations: 90.656 KiB)
PagedMergeSort, 8 threads:    0.223406 seconds (212 allocations: 746.016 KiB)

sortperm! T=UInt64 n=8_388_608 (64.0 MiB)
default:                      5.376910 seconds (2 allocations: 64.000 MiB, 0.01% gc time)
MergeSort:                    1.776138 seconds (2 allocations: 32.000 MiB)
PagedMergeSort:               1.772253 seconds (4 allocations: 90.656 KiB)
PagedMergeSort, 8 threads:    0.673832 seconds (208 allocations: 746.375 KiB)

sort! T=UInt64 n=134_217_728 (1.0 GiB)
default:                      7.397737 seconds (3 allocations: 1.000 GiB, 0.05% gc time)
MergeSort:                   19.119953 seconds (2 allocations: 512.000 MiB, 0.67% gc time)
PagedMergeSort:              15.683231 seconds (4 allocations: 362.219 KiB)
PagedMergeSort, 8 threads:    4.070665 seconds (212 allocations: 2.850 MiB)

sortperm! T=UInt64 n=134_217_728 (1.0 GiB)
default:                    176.440430 seconds (2 allocations: 1.000 GiB, 0.07% gc time)
MergeSort:                   41.310611 seconds (2 allocations: 512.000 MiB, 0.59% gc time)
PagedMergeSort:              40.876463 seconds (4 allocations: 362.219 KiB)
PagedMergeSort, 8 threads:   16.644219 seconds (214 allocations: 2.851 MiB)
```

```
Julia Version 1.9.0-beta3
Commit 24204a7344 (2023-01-18 07:20 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 24 × AMD Ryzen 9 3900X 12-Core Processor            
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, znver2)
  Threads: 12 on 24 virtual cores
Environment:
  JULIA_DEPOT_PATH = C:\Users\Luke\.julia;C:\Users\Luke\.julia\juliaup\julia-1.9.0-beta3+0.x64.w64.mingw32\local\share\julia;C:\Users\Luke\.julia\juliaup\julia-1.9.0-beta3+0.x64.w64.mingw32\share\julia
  JULIA_IMAGE_THREADS = 1
  JULIA_LOAD_PATH = @;@v#.#;@stdlib
  JULIA_REVISE_WORKER_ONLY = 1

sort! T=UInt64 n=131_072 (1.0 MiB)
default:                      0.002453 seconds (3 allocations: 1.008 MiB)
MergeSort:                    0.007748 seconds (2 allocations: 512.047 KiB)
PagedMergeSort:               0.006455 seconds (2 allocations: 11.625 KiB)
PagedMergeSort, 12 threads:   0.001617 seconds (134 allocations: 150.641 KiB)

sortperm! T=UInt64 n=131_072 (1.0 MiB)
default:                      0.005796 seconds (2 allocations: 1.000 MiB)
MergeSort:                    0.009661 seconds (2 allocations: 512.047 KiB)
PagedMergeSort:               0.009557 seconds (2 allocations: 11.625 KiB)
PagedMergeSort, 12 threads:   0.002576 seconds (130 allocations: 150.750 KiB)

sort! T=UInt64 n=8_388_608 (64.0 MiB)
default:                      0.274773 seconds (3 allocations: 64.008 MiB)
MergeSort:                    0.693271 seconds (2 allocations: 32.000 MiB)
PagedMergeSort:               0.558476 seconds (4 allocations: 90.656 KiB)
PagedMergeSort, 12 threads:   0.106098 seconds (393 allocations: 1.102 MiB)

sortperm! T=UInt64 n=8_388_608 (64.0 MiB)
default:                      1.129249 seconds (2 allocations: 64.000 MiB)
MergeSort:                    1.175415 seconds (2 allocations: 32.000 MiB)
PagedMergeSort:               1.118836 seconds (4 allocations: 90.656 KiB)
PagedMergeSort, 12 threads:   0.369979 seconds (394 allocations: 1.103 MiB)

sort! T=UInt64 n=134_217_728 (1.0 GiB)
default:                      6.363665 seconds (3 allocations: 1.000 GiB, 0.01% gc time)
MergeSort:                   13.181063 seconds (2 allocations: 512.000 MiB)
PagedMergeSort:              10.717851 seconds (4 allocations: 362.219 KiB)
PagedMergeSort, 12 threads:   1.958344 seconds (401 allocations: 4.285 MiB)

sortperm! T=UInt64 n=134_217_728 (1.0 GiB)
default:                     34.260657 seconds (2 allocations: 1.000 GiB, 0.00% gc time)
MergeSort:                   29.027326 seconds (2 allocations: 512.000 MiB)
PagedMergeSort:              28.057972 seconds (4 allocations: 362.219 KiB)
PagedMergeSort, 12 threads:   7.543024 seconds (394 allocations: 4.286 MiB)
```

LilithHafner commented:
Thank you! This is a valuable contribution! I hope you don't mind some critique :)

It's a very large sorting algorithm so I have yet to read it cover-to-cover. If possible, it would be nice to get similar or better performance with a simpler implementation, but that can be tricky. That said, I have performed some black-box analysis with a couple results:

<details>

<summary>First, a correctness issue, I ran this script to check stability and found a counterexample:</summary>

```julia
@testset begin
    for T in (Float64, Int, UInt8)
        for alg in [ThreadedPagedMergeSort, PagedMergeSort]
            for order in [Forward, Reverse, By(identity), By(abs), By(Returns(0)), By(Base.Fix2(÷, 100))]
                fails = Int[]
                for n in vcat(0:30, 40:10:100, 110:50:1000)
                    v = rand(T, n)
                    # @test sort(v; order) == sort(v; alg, order)
                    sort(v; order) == sort(v; alg, order) || push!(fails, n)
                end
                if !isempty(fails)
                    println("$(T) $(alg) $(typeof(order))\n\t$(fails)")
                end
            end
        end
    end
end
```

</details>

```julia
julia> issorted(sort(1:100, by=Returns(0), alg=PagedMergeSort))
false
```

Second, some notes on benchmarking, use `BenchmarkTools.@btime` rather than `@time` for more accurate results. `@btime` executes the test expression repeatedly and takes the minimum to reduce noise. I also prefer to work with benchmarking code that can be pasted into a REPL because that is more widely and quickly reproducible than a Pluto notebook. For example, the following can be `include`ed or pasted into a REPL

```julia
using BenchmarkTools, SortingAlgorithms, Random, Test
versioninfo()
for i in 0:5
    n = 17^i
    println("sort!(rand(Int, $n))")
    v = rand(Int, n)
    print("Default:  "); @btime sort!($v) setup=(rand!($v)) evals=1
    print("Paged MS: "); @btime sort!($v; alg=PagedMergeSort) setup=(rand!($v)) evals=1
    print("Threaded: "); @btime sort!($v; alg=ThreadedPagedMergeSort) setup=(rand!($v)) evals=1
end
```

<details>

<summary>On my computer, it produces the following results:</summary>

```julia
Julia Version 1.9.0-beta4
Commit b75ddb787ff (2023-02-07 21:53 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin21.4.0)
  CPU: 4 × Intel(R) Core(TM) i5-8210Y CPU @ 1.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, skylake)
  Threads: 2 on 2 virtual cores

sort!(rand(Int, 1))
Default:    63.000 ns (0 allocations: 0 bytes)
Paged MS:   64.000 ns (0 allocations: 0 bytes)
Threaded:   63.000 ns (0 allocations: 0 bytes)
sort!(rand(Int, 17))
Default:    248.000 ns (0 allocations: 0 bytes)
Paged MS:   390.000 ns (2 allocations: 256 bytes)
Threaded:   386.000 ns (2 allocations: 256 bytes)
sort!(rand(Int, 289))
Default:    5.214 μs (1 allocation: 2.44 KiB)
Paged MS:   8.231 μs (2 allocations: 704 bytes)
Threaded:   8.453 μs (2 allocations: 704 bytes)
sort!(rand(Int, 4913))
Default:    91.831 μs (3 allocations: 46.67 KiB)
Paged MS:   206.361 μs (2 allocations: 2.38 KiB)
Threaded:   203.543 μs (2 allocations: 2.38 KiB)
sort!(rand(Int, 83521))
Default:    1.757 ms (3 allocations: 660.80 KiB)
Paged MS:   4.627 ms (2 allocations: 9.38 KiB)
Threaded:   3.074 ms (54 allocations: 23.03 KiB)
sort!(rand(Int, 1419857))
Default:    39.456 ms (3 allocations: 10.84 MiB)
Paged MS:   111.730 ms (3 allocations: 37.48 KiB)
Threaded:   69.457 ms (57 allocations: 79.28 KiB)
```

</details>

LilithHafner reviewed (commented):
Here are a few inline comments from looking over the parts of the code I understand. I'll need to get a better understanding of block sorts before I can give a full review.

LSchwerdt commented:
Thank you for taking the time to review this PR. I am looking forward to the critique :).

> First, a correctness issue, I ran this script to check stability and found a counterexample:

That was caused by the initial optimizations. Base sort returned a range, while all algorithms from SortingAlgorithms returned a vector. Fixed now after implementing initial optimizations.

> Second, some notes on benchmarking, use `BenchmarkTools.@btime` rather than `@time` for more accurate results.

This algorithm is most useful for sorting large amounts of data, where `@btime` uses only one sample anyways. I only included the smaller input sizes to show the memory usage (allocations). But you are right, `@btime` would have avoided doubts about the timing accuracy.

> I also prefer to work with benchmarking code that can be pasted into a REPL

I found Pluto notebooks that load the correct package from github very convenient when benchmarking on different PCs not setup for package development. Run the notebook -> done. But you are right, if you already checked out the branch for review, a simple script is even more convenient. I have added a benchmark script to PR #72.

> it would be nice to get similar or better performance with a simpler implementation

There are some ways to simplify the code, but all are less performant:

- Use array assignments instead of loops.
- Use a reshaped view of the input array (blocksize x nBlocks matrix). The last elements that do not fit would need to be handled seperately, though.
- Dont use twoended_merge!

Feel free to suggest improvements in this regard. I will continue trying to simplify, too. But this is the main reason I did not go for an algorithm with O(1) space. While not especially elegant, this PR is much shorter and concenpually simpler than [HolyGrailSort](https://github.com/HolyGrailSortProject/Holy-Grailsort/blob/main/Holy%20Grail%20Sort/Java/Summer%20Dragonfly%20et%20al.'s%20Rough%20Draft/src/holygrail/HolyGrailSort.java), for example.

LilithHafner commented:
> Fixed now after implementing initial optimizations.

I'm still seeing a problem:

```julia
julia> PagedMergeSort
Base.Sort.MissingOptimization(
    Base.Sort.BoolOptimization(
        Base.Sort.Small{10}(
            Base.Sort.InsertionSortAlg(),
            Base.Sort.IEEEFloatOptimization(
                SortingAlgorithms.PagedMergeSortAlg()))))

julia> sort(1:100, by=Returns(0), alg=PagedMergeSort)
100-element Vector{Int64}:
   1
   2
   3
   4
   5
   6
   7
  14
  15
  16
   ⋮
  98
  99
 100
  83
  84
  85
  86
  87
  88
```

LSchwerdt commented:
> I'll need to get a better understanding of block sorts before I can give a full review.

This is not a [block sort as defined here](https://en.wikipedia.org/wiki/Block_sort). That's why I called it PagedMergeSort and not BlockMergeSort, although block merge sort would be a fitting name.

Btw, is there a good way to make the gif in the OP available in the documentation somehow? I think it is very helpful to understand the merge procedure.

LSchwerdt commented on src/SortingAlgorithms.jl:
```diff
@@ -626,4 +675,399 @@ else
+                iAL += 1
+            end
+            oL += 1
+        end
```
Not according to Codecov ;)

This runs when either A or B only contain one element, as the two-ended merge could take 2 elements from either A or B.

Added clarifying comments in [dad2a2c](https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/71/commits/dad2a2cc8bf9a1eb9d5459b0c0e60e84a7045966).

LilithHafner replied:
Right! I missed that the above loop is based on < and this one is <=.

LSchwerdt commented:
> I'm still seeing a problem:

Sorry, my bad. I misread. With [dab157f](https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/71/commits/dab157f7d7962f91c374f08f25056116b8dab0da) the whole testset should complete without errors now. (It does for me.)

LSchwerdt commented on src/SortingAlgorithms.jl:
```diff
@@ -626,4 +675,399 @@ else
+# -> redefine for compatibility with earlier versions
+_midpoint(lo::Integer, hi::Integer) = lo + ((hi - lo) >>> 0x01)
+
+function pagedmergesort!(v::AbstractVector{T}, lo::Integer, hi::Integer, buf::AbstractVector{T}, blockLocation, o=Base.Order.Forward) where T
```
Fixed in [ea4eb9a](https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/71/commits/ea4eb9ad4fe192e3c24f296be4e4786b91480cbb).

I looked too much at radix sort, where `bits` and `t` are right after `hi`. But I didnt notice that there is no Ordering argument...

LSchwerdt commented on src/SortingAlgorithms.jl:
```diff
@@ -626,4 +675,399 @@ else
+    oR = 1 + hi - lo
+
+    # input array indices
+    iAL = loA
+    iBL = loB
+    iAR = hiA
+    iBR = hiB
```
Could you clarify that? I don't think I understand exactly what you mean. I use A and B to avoid confusion between the left subarray and the left index. lo/hi are used to specify the input size/location, not as running indices. Here is a table with the variable names:

| twoended_merge! | merge! | MergeSort (Base.Sort) | Description | 
| ------------- | ------------- | ------------- | ------------- |
| lo | lo  | lo | first index of the input |
| hi | hi | hi | last index of the input |
| hiA |  | m | last index of the left subarray |
| iAL | iA | i | current left index of the left input subarray |
| iAR |  |  | current right index of the left input subarray |
| iBL | iB | j | current left index of the right input subarray |
| iBR |  |  | current right index of the right input subarray |
| oL | k | k | current left index of the output |
| oR |  |  | current right index of the output |

LilithHafner replied:
The fact that `L` means `lo` but a running index and `R` means `hi` but a running index is unintuitive to me. I don't like switching from one synonym or metaphor to another to mean something unrelated. I see no connection between the low/high metaphor and static variables nor between the left/right metaphor and mutable variables.

As an alternative,

| twoended_merge! | merge! | MergeSort (Base.Sort) | Description | 
| ------------- | ------------- | ------------- | ------------- |
| lo | lo  | lo | first index of the input |
| hi | hi | hi | last index of the input |
| m |  | m | last index of the left subarray |
| a_lo | a | a | current left index of the left input subarray |
| a_hi |  |  | current right index of the left input subarray |
| b_lo | b | b | current left index of the right input subarray |
| b_hi |  |  | current right index of the right input subarray |
| o_lo | k | k | current left index of the output |
| o_hi |  |  | current right index of the output |

I like this because it has one fewer component (lo/hi and a/b/lo rather than lo/hi, iA/iB/o, and L/R); because mutable variables and static variables are more clearly distinguishable (unlike imo `hiA` and `iAR` which blend together); and because the call signature variable names (`lo`/`hi`/`m`) make sense as a triple without needing to dive into the mutable variable name conventions (imo unlike `lo`/`hi`/`hiA`)

LSchwerdt replied:
I used L and R, because in some papers on Quicksort the pointers are called left and right pointers (L, R). And somehow the mutable working indices and the static limits lo/hi were conceptually not the same for me.

But code must be easily understood by people who did not write it, so your opinion is more important. Besides, I like your suggestion and implemented it in [7b2a653](https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/71/commits/7b2a653b24877d464bef9d9d57649a3011050985) with one small change:
`o_hi` is visually easy to confuse with the ordering `o`. And since `i` for input indices is not used anymore, using `o` is not really necessary for consistency, so I use `k` for the output indices everywhere.

LilithHafner replied:
I like that change. `o` is a bad variable name for integers anyways :)

LilithHafner commented:
Funny story, I saw that 2016 tests passed and I thought "oh, that's cute. That's a year. Let's make it in the future" and changed
```julia
for n in vcat(0:30, 40:10:100, 110:50:1000)
```
to
```julia
for n in vcat(0:31, 40:10:100, 110:50:1000)
```
revealing this error:
```julia
julia> sort(1:31; alg=PagedMergeSort)
ERROR: not implemented
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:35
  [2] pagedMerge!(v::Vector{Int64}, lo::Int64, hiA::Int64, hi::Int64, o::ForwardOrdering, buf::Vector{Int64}, blockLocation::Vector{Int64})
    @ SortingAlgorithms ~/.julia/dev/SortingAlgorithms/src/SortingAlgorithms.jl:823
  [3] pagedmergesort!(v::Vector{Int64}, lo::Int64, hi::Int64, o::ForwardOrdering, buf::Vector{Int64}, blockLocation::Vector{Int64})
    @ SortingAlgorithms ~/.julia/dev/SortingAlgorithms/src/SortingAlgorithms.jl:1012
  [4] sort!(v::Vector{Int64}, lo::Int64, hi::Int64, #unused#::SortingAlgorithms.PagedMergeSortAlg, o::ForwardOrdering)
    @ SortingAlgorithms ~/.julia/dev/SortingAlgorithms/src/SortingAlgorithms.jl:1024
  [5] _sort!
    @ ./sort.jl:2135 [inlined]
  [6] _sort!
    @ ./sort.jl:643 [inlined]
  [7] _sort!
    @ ./sort.jl:714 [inlined]
  [8] _sort!
    @ ./sort.jl:659 [inlined]
  [9] _sort!
    @ ./sort.jl:595 [inlined]
 [10] #sort!#28
    @ ./sort.jl:1373 [inlined]
 [11] #sort#29
    @ ./sort.jl:1399 [inlined]
 [12] kwcall(::NamedTuple{(:alg,), Tuple{Base.Sort.MissingOptimization{Base.Sort.BoolOptimization{Base.Sort.Small{10, Base.Sort.InsertionSortAlg, Base.Sort.IEEEFloatOptimization{SortingAlgorithms.PagedMergeSortAlg}}}}}}, ::typeof(sort), v::UnitRange{Int64})
    @ Base.Sort ./sort.jl:1399
 [13] top-level scope
    @ REPL[14]:1
```
The lesson here is that exhaustive end-to-end testing is hard for large algorithms.

LilithHafner commented:
> This is not a [block sort as defined here](https://en.wikipedia.org/wiki/Block_sort). That's why I called it PagedMergeSort and not BlockMergeSort, although block merge sort would be a fitting name.

lol okay. Oops.

> Btw, is there a good way to make the gif in the OP available in the documentation somehow? I think it is very helpful to understand the merge procedure.

Perhaps this will help: https://discourse.julialang.org/t/including-generated-gif-files-in-the-documentation/17363/5

LSchwerdt commented:
> The lesson here is that exhaustive end-to-end testing is hard for large algorithms.

Good catch! I specifically remember testing this, but obviously I must have messed up. Fixed by [d9fcaef](https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/71/commits/d9fcaef78955f01e0d79ec76f8e93c8d1615ede4).
I think, this is a good opportunity to improve the test set and added that to the todo list in the OP.

LSchwerdt commented:
> Perhaps this will help: https://discourse.julialang.org/t/including-generated-gif-files-in-the-documentation/17363/5

Thank you, but I think this only works when using Documenter.jl to generate html documentation. Using ? in the repl or viewing the docstring in Pluto does not include or link the gif. I have added it anyways in the docs folder, in case someone looks at the package source to learn how the algorithm works. Or did I miss something?

LilithHafner commented on src/SortingAlgorithms.jl:
```diff
@@ -631,4 +684,376 @@ else
+            @getNextBlock!
+            k_block = 1
+            offset = getBlockOffset(currentBlock)
+            while k_block <= blocksize && a <= m && b <= hi
```
IIUC once A is within one block of empty, there will be fewer than 3 empty blocks left in A, so there is room to merge everything into the end of B. Similarly, once B is within one block of empty it is sound to merge blindly into A until one of A or B runs out. It's possible to do even better still in the second case by switching to merging into the end of B to reduce swapping time in the next phase, but that may not be worth the complexity.

You would have to modify block management slightly to accomadate this, but it shouldn't be too bad.

LSchwerdt replied:
There may be special cases, where the check is not necessary, but I dont think this is true in general. Doing it the way it is currently done, definitively results in simpler code.

In this frame from the example gif you can see that both A and B are within one block of empty when starting to merge into a new block, but there is not enough space to merge into A. Depending on which blocks were used previously, there might not be enough space in B.
![pagedMerge_130_130_finalmerge](https://user-images.githubusercontent.com/8214073/226172313-7f0aab18-65e3-4614-b5a3-30bdd04680f5.png)

LilithHafner reviewed (commented):
You write the standard merge routine four times with different end conditions. Perhaps a higher order function would be more readable
```julia
@propagate_inbounds function merge!(f,target,source,a,b,k)
    while f(a,b,k)
        if lt(o, source[b], source[a])
            target[k] = source[b]
            b += 1
        else
            target[k] = source[a]
            a += 1
        end
        k += 1
    end
    a,b,k
end

...

a,b,k = merge!(buf, v, a, b, 1) do a,b,k
    k <= 3blocksize
end
```

> I am not 100% happy with using a macro (getNextBlock!) instead of a function, but I could not come up with a better way.

I think a function which takes a bunch of integer parameters and returns their new values is more readable than a macro because the parameters lists, while potentially unwieldily, are at least present. e.g.

```julia
function getNextBlock!(blockLocation, currentBlockIdx, nextBlockA, nextBlockB, a, b, blocksize)
    if a > nextBlockA * blocksize + lo
        currentBlock = nextBlockA
        nextBlockA += 1
    else
        currentBlock = nextBlockB
        nextBlockB += 1
    end
    blockLocation[currentBlockIdx] = currentBlock
    currentBlockIdx + 1, nextBlockA, nextBlockB
end
```

I suspect it is possible to improve the readability further, still.

LilithHafner commented:
I think that the process of moving the pages into sorted order can be simpler: loop over the three free pages at the end of the subarrays and for each, move the page that belongs there into position, and then the page that belongs where that page came from, so on and so forth until moving a page out of the buffer into position in one if the initial blocks of one of the two subarrays.

Once this is done, we know that the buffer is free and all that remains are closed cycles of pages in the output vector. Search linearly until finding an out of position page, copy that page to the first entry in the buffer, permute the cycle into position, and  copy the initial page back from the buffer into its final position. Continue the linear search from where we left off.

(Sadly, the linear scan is actually required in some cases. For example, if the initial vector is 
`[4,5,6,7,8,9,10,1,2,3,11,12,13]` and the block size is 1 (or larger and all blocks consist of the same number repeated several times), then the vector after the initial merge will be `[5,6,7,8,9,10,11,4,12,13,?,?,?]` and after the initial pass will be `[1,2,3,8,5,6,7,4,9,10,11,12,13]`. Credit to Zion Hefty for the example.)

This is maximally efficient (exactly one page is copied to a location other than its final destination for each closed cycle of pages that need to be copied), simpler than the current implementation (e.g. don't need to keep track of an array of all three currently free page positions), and drops the dependency on StaticArrays.

<!--I think that the process of moving the pages into sorted order can be much simpler and slightly more efficient.

The key idea is that the page table is free of non-trivial cycles (i.e. cycles with length more than 1) Proof*

From this fact, we can iterate over the free blocks (there are exactly three) and for each free block, trace that block's inverse permutation.

Once this is complete, the page table is still free of non-trivial cycles, and is now a permutation, and is therefore the identity permutation, and therefore sorted.

Pseudocode:
```julia
for i in free_blocks
    while [i refers to an block in the output vector rather than the buffer]
        j = page_table[i]
        copyto!(v, i, v, j)
        i = j
    end
end
```

This is nice because it is maximally efficient (every copy operation takes a chunk to its final position; chunks that are already in their final position are not touched), simple, and drops the StaticArrays dependancy.

*Proof:

Assume to reach a contradiction that after the merge step and before the re-arrange block step there is a nontrivial cycle of (a,b,c,...,a) where the block a belongs in the position b is in, the block b belongs in the position c is in, and so on an so forth. Because this cycle is nontrivial, it has at least two elements. Let's consider the first two, a and b. We know that a belongs in the position b is in, and there is a chain of belonging where b belongs in the position another block is in which belongs in a position a is in (though that number of "another block"s could by 0 or any other non-negative integer).

Breaking this into cases, assume a is in the first half of the array (i.e. in A). 

...

Consider a page _a_ in B that belongs elsewhere in B. _a_ must belong in a position after its current position. [Proof TODO]
Consider a page _a_ in A that belongs elsewhere in A. _a_ must belong in a position after its current position. [Proof TODO]-->

LSchwerdt commented:
> I think that the process of moving the pages into sorted order can be simpler: [...]

If I understand it correctly, what you describe is in essence the current solution. But I will try to improve it to make this obvious when looking at the code. And maybe I'll manage to remove the dependency on StaticArrays, by making the control flow explicit, following your suggestion.

Looping through the three free pages at the end of the subarrays is accomplished using `freeBlocksIdx`. It starts at 3, and after the 3 pages are moved from the buffer, only the first page in the buffer is used for all the following permutation cycles.

The linear scan is tracked with `doneBlockIdx`. It is initialized to 1, and `while true` is exited by `doneBlockIdx == nBlocks && return` when the linear scan is complete.

LilithHafner commented:
Great work! It is much simpler now. I think it can be even simpler, though. I've put some suggestions in a [pr to the head branch of this pr](https://github.com/LSchwerdt/SortingAlgorithms.jl/pull/1) because I wanted to make sure that my suggestions worked before recommending them.

LilithHafner commented on src/SortingAlgorithms.jl:
```diff
@@ -631,4 +683,308 @@ else
+                pages = next_page!(pageLocations, pages, pagesize, lo, a)
+                offset = page_offset(pages.current)
+                copyto!(v, offset+1, v, use_a ? a : b, pagesize)
+                use_a && (a += pagesize)
+                use_a || (b += pagesize)
```
This is funny and a little off topic: I prefer each style in this list over the one immediately before it
```julia
use_a && (a += pagesize)
use_a || (b += pagesize)
```
```julia
use_a ? (a += pagesize) : (b += pagesize)
```
```julia
if use_a (a += pagesize) else (b += pagesize) end
```
```julia
if use_a
    a += pagesize
else 
    b += pagesize
end
```
```julia
use_a && (a += pagesize)
use_a || (b += pagesize)
```
(yes, that is a circular preference loop, it's probably fine as is)

LSchwerdt replied:
I agree.

LilithHafner commented on src/SortingAlgorithms.jl:
```diff
@@ -631,4 +683,308 @@ else
+        # initialize variables for merging into pages
+        pages = Pages(-17, 0, 1, (m-lo) ÷ pagesize + 2) # first argument is unused
+        # more efficient loop while more than pagesize elements of A and B are remaining
+        while_condition1(offset) = (_,_,k) -> k <= offset + pagesize
+        while a < m-pagesize && b < hi-pagesize
+            pages = next_page!(pageLocations, pages, pagesize, lo, a)
+            offset = page_offset(pages.current)
+            a,b,_ = merge!(while_condition1(offset),v,v,v,o,a,b,offset+1)
+        end
+        # merge until either A or B is empty or the last page is reached
+        k, offset = nothing, nothing
+        while_condition2(offset) = (a,b,k) -> k <= offset + pagesize && a <= m && b <= hi
+        while a <= m && b <= hi && pages.currentNumber + 3 < nPages
+            pages = next_page!(pageLocations, pages, pagesize, lo, a)
+            offset = page_offset(pages.current)
+            a,b,k = merge!(while_condition2(offset),v,v,v,o,a,b,offset+1)
+        end
```
These while loops and `while_condition`s feel strange to me; I wonder if there is a better approach.

LSchwerdt replied:
It was the best way to implement this I could think of. But of course I am open to suggestions for implrovement.

LilithHafner commented on src/SortingAlgorithms.jl:
```diff
@@ -631,4 +683,308 @@ else
+        put!(c, i)
+    end
+    threaded_pagedmergesort!(v, lo, hi, o, bufs, pageLocations, c, threadingThreshold)
+    return v
+end
+else
+    # use single threaded function when VERSION < v"1.3"
```
This indentation is inconsistent

LilithHafner reviewed (commented):
> Rename buffer/buf to scratch or to t?
Base.Sort's "scratch" is an implementation detail and will likely change semantics to use [buffers](https://gist.github.com/JeffBezanson/a25dde3bebb5a734af87bb5ddcf31fb0) once they are a thing, so it's for the best not to integrate too closely with that system.

However, I still think sharing a common name might be a good idea. IIRC, Base uses `t` and `scratch`, the (obsolete) radix sort in SotihngAlgorithsm.jl uses `ts`.

Right now, PagedMergeSort uses both `t` and `buf` which brings the total number of names for this concept up to 3 in this repo and 4 in this repo+base. I recommend using `scratch` as an abbreviation for "scratch space".

> CI only tests the fallback to PagedMergeSort for ThreadedPagedMergeSort.
That's not good. Maybe we should split off PagedMergeSort and try to merge it first and separately, and do ThreadedPagedMergeSort in a second PR.

LSchwerdt commented on src/SortingAlgorithms.jl:
```diff
@@ -631,4 +683,308 @@ else
+    len = hi + 1 -lo
+    if len <= Base.SMALL_THRESHOLD
+        return Base.Sort.sort!(v, lo, hi, Base.Sort.InsertionSortAlg(), o)
+    end
```
No longer needed due to maybe_optimize.

LSchwerdt commented:
> Right now, PagedMergeSort uses both `t` and `buf` which brings the total number of names for this concept up to 3 in this repo and 4 in this repo+base. I recommend using `scratch` as an abbreviation for "scratch space".

> That's not good. Maybe we should split off PagedMergeSort and try to merge it first and separately, and do ThreadedPagedMergeSort in a second PR.


These are good suggestions imho. I have implemented them in [3f337f0](https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/71/commits/3f337f04a47cef461681dd2fe805bc746f3b3f65) and [d3ea719](https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/71/commits/d3ea719ad618f0ca6de112b1316ea7beb1bffaf5).

Except for rebasing, my checklist of open questions is now completed.

LilithHafner commented on src/SortingAlgorithms.jl:
```diff
@@ -652,4 +681,239 @@ else
+            use_a = a <= m
+            # copy the incomplete page
+            partial_page_size = offset + pagesize - k + 1
+            copyto!(v, k, v, use_a ? a : b, partial_page_size)
```
It turns out I was wrong when I said it was bad to reinvent the wheel with `copyto!`. `Base.copyto!` is [buggy](https://github.com/JuliaLang/julia/issues/50900). Because it tries to detect aliasing which is not really possible to do effectively in Julia.

We should roll our own :`(
```julia
function _unsafe_copyto!(dest, doffs, src, soffs, n)
    @inbounds for i in 0:n-1
        dest[doffs + i] = src[soffs + i]
    end
    dest
end

function _unsafe_copyto!(dest::Array, doffs, src::Array, soffs, n)
    unsafe_copyto!(dest, doffs, src, soffs, n)
end
```

LSchwerdt replied:
> It turns out I was wrong when I said it was bad to reinvent the wheel [...].

Good catch! But beeing right twice does not mean you were wrong. :wink: Using copyto! definitively led to better code. Even after using the custom unsafe version.

LilithHafner commented on src/SortingAlgorithms.jl:
```diff
@@ -652,4 +681,239 @@ else
+    pagesize = isqrt(n)
+    scratch = Vector{eltype(v)}(undef, 3pagesize)
+    nPages = n ÷ pagesize
+    pageLocations = Vector{Int}(undef, nPages - 3)
```
If `nPages < 3`, this will throw. This happens when `2 <= length(lo:hi) <= 5`.

LSchwerdt replied:
Fixed in [717f9ae](https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/71/commits/717f9ae6273c3e14136354eae3c4cc4c09e26629).
Do you think an extra test is needed for this?
(i.e. test SortingAlgorithms.PagedMergeSortAlg() in addition to PagedMergeSort)

LilithHafner replied:
CI revealed this bug, so it's not totally necessary, but it certainly wouldn't hurt.

LSchwerdt replied:
Test implemented in [d367ee8](https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/71/commits/d367ee8a17a087c6c54994e9375269782e8334dd).

LilithHafner reviewed (approved):
LGTM!

LSchwerdt commented:
> Thank you! This is a nice algorithm to have.

Thank you for your efforts reviewing and massively improving the code.
